### PR TITLE
fix(composer): Handle minimum-stability version constraints

### LINKED ISSUE #6108: Certain version constraints for composer are not correctly handled
**What Renovate type are you using?**

It's a hosted app on GitLab.

**Describe the bug**

This bug is as a result of #6057. During updates, it can generate version constraints which cannot be read by composer. This happens under these conditions:

1. The package's version is a pre-release version (such as alpha, beta, etc)
2. There is a specific pre-release sequence suffix (such as `beta2`, `rc1`)

This bug doesn't break any PRs which would be generated before #6057 was merged unless `packageRules` config causes such packages to be grouped with other regular packages.

**Did you see anything helpful in debug logs?**

This is the composer command that failed and its output:

```
"artifactErrors": [
  {
    "lockFile": "composer.lock",
    "stderr": "Command failed: composer update drupal/anchor_link drupal/better_exposed_filters drupal/blocktabs drupal/cdn drupal/copyright_block drupal/cshs drupal/disqus drupal/editor_advanced_link drupal/entity_browser drupal/entity_browser_block drupal/facebook_pixel drupal/hubspot_embed drupal/multiple_fields_remove_button drupal/paragraphs drupal/path_redirect_import drupal/redis drupal/s3fs drupal/search_api_solr drupal/slick_views drupal/twig_tweak drupal/views_data_export drupal/views_slideshow --with-dependencies --ignore-platform-reqs --no-ansi --no-interaction --no-scripts --no-autoloader\n\n                                                                                          \n  [UnexpectedValueException]                                                              \n  Could not parse version constraint ^4.0.0@beta1: Invalid version string \"^4.0.0@beta1\"  \n                                                                                          \n\n"
  }
]
```

The concerned package:
```json
{
 "depType": "require",
 "depName": "drupal/better_exposed_filters",
 "currentValue": "^4.0@alpha",
 "datasource": "packagist",
 "lockedVersion": "4.0.0-alpha1",
 "depIndex": 19,
 "updates": [
   {
     "fromVersion": "4.0.0-alpha1",
     "toVersion": "4.0.0-beta1",
     "newValue": "^4.0.0@beta1",
     "newMajor": 4,
     "newMinor": 0,
     "updateType": "minor",
     "isSingleVersion": false,
     "isRange": true,
     "isBump": true
   }
 ],
 "warnings": [],
 "sourceUrl": "https://git.drupalcode.org/project/better_exposed_filters.git",
 "homepage": "https://www.drupal.org/project/better_exposed_filters"
},
```

I'll work on a PR in a day or two.

**To Reproduce**

**Additional context**

### PR DESCRIPTION
**Pull request by hussainweb**
Closes #6108

### CODE REVIEW

**Comment by hussainweb** [comment]
All expected failures. I'm about to push a tentative fix for this now.

**Comment by JamieMagee** [comment]
@hussainweb Looks fine, but do you have a sample PR you can show?

File: lib/versioning/composer/index.spec.ts [hunk 1]
```diff
@@ -55,9 +55,14 @@ describe('semver.isValid(input)', () => {
     expect(semver.isValid('>1.2.3')).toBeTruthy();
   });
   it('should support ranges with stability modifiers', () => {
-    expect(semver.isValid('~1.2.3@beta1')).toBeTruthy();
```
[the remark this answers is not in the document]
Comment (hussainweb) [reply lib/versioning/composer/index.spec.ts:58]: Yes, this is not a valid composer constraint. I added this in my previous PR and I was mistaken about this.

**Comment by hussainweb** [comment]
> @hussainweb Looks fine, but do you have a sample PR you can show?

@JamieMagee I'm not sure which PR you mean. Do you mean a PR/MR that renovate creates? There's one in #6108. Please let me know if you need more details about that MR.

**Comment by rarkins** [comment]
@hussainweb I think he means a demo public repo you have tested with. i.e. can you show this working somewhere?

**Comment by hussainweb** [comment]
@rarkins, I'm afraid not. Is there a guide I can follow to quickly test this? I am happy to create a repo with sample composer.json files but I haven't run patch versions of Renovate before to actually test the fix.

**Comment by rarkins** [comment]
@hussainweb for this type of PR it's best to run it against a "real" repository instead of only unit tests. Here's how you can do it, e.g. on github.com:

1. Create the repo (e.g. github.com/hussainweb/composer1) on github.com, add composer.json/composer.lock
2. Add an empty `renovate.json` (`{}`)
3. From your forked repo clone, run `yarn start hussainweb/composer1`). You will need to make sure you are passing your token to it somehow, either in env, CLI or config.js

**Comment by JamieMagee** [comment]
@hussainweb Yes, I'd like to run it against a real repository. If you can point me to a repository with a composer setup, I can fork your repo and run this branch of Renovate against it.

**Comment by hussainweb** [comment]
Thanks @rarkins for the setup instructions and @JamieMagee for the offer to run it. I have created a repository with a composer setup here: https://github.com/hussainweb/composer-renovate-test

I ran Renovate from my fork as per @rarkins instructions and it generated a PR: https://github.com/hussainweb/composer-renovate-test/pull/1. The PR failed, but that's because of a composer error. The version number replacements seem to work as expected.

**Comment by hussainweb** [comment]
@rarkins @JamieMagee, BTW, the composer memory issue is typical when working with Drupal. We usually work around it by increasing the memory limit for composer by setting an environment variable `COMPOSER_MEMORY_LIMIT=-1`. Any suggestions on how to handle it here? I recognise this has to be a different issue/PR but I am just looking for ideas.

**Comment by JamieMagee** [comment]
@hussainweb Thanks, this looks good. I'll update this branch and merge once all tests pass.

As for setting environment variables, it should be easy. We can just add another environment variable [here](https://github.com/renovatebot/renovate/blob/master/lib/manager/composer/artifacts.ts#L111-L113). I'm a little uneasy about giving composer unlimited memory by default, but we can certainly increase it.

**Review by JamieMagee — APPROVED** [review]

**Pull request merged**